### PR TITLE
Task/improve compose screen components

### PR DESCRIPTION
**Pull request by filbabic**
### 🎯 Goal

It was confusing to use Screen components and their respective ViewModels(or factories) and expect results in both to be the same. This could and couldn't be true because the Screen component created its own ViewModels and factories, and the way the user could create their own could be different.

Additionally, it was tedious to send customization params for behavior through the local viewModelFactory and the screen component, because you could customize something in one place but not in another.

This way, we're "forcing" users to pass in ViewModelFactories and as such, the ViewModels used outside the components and inside should always be the same (as they're cached in the factory/provider). Additionally, you only customize the behavior params (like message limits and similar) in one place (the factory), so the screen components have fewer parameters, are easier and cleaner to use.

Finally, our documentation is up to date and shows how to override ViewModels and use the same ones in both the inner components and the outside world.

### 🛠 Implementation details

Made the ViewModelFactories mandatory params of screen components, but also set up default values for their constructors, so it's much easier to create and pass them in.

Outlined important parameters for both the Messages and Channels screens.

Updated all the docs and examples.

### 🎨 UI Changes

N/A - there shouldn't be any changes in the UI or behavior. 

### 🧪 Testing

To test this, apply the following patch and search the Logcat console for `ViewModelAddress`:

```
ViewModelAddress: ChannelListViewModel@e86d6c8
ViewModelAddress: ChannelListViewModel@e86d6c8
ViewModelAddress: ChannelListViewModel@e86d6c8
ViewModelAddress: MessageListViewModel@5eb7a75
ViewModelAddress: AttachmentsPickerViewModel@10fa0a
ViewModelAddress: MessageComposerViewModel@f90a898
ViewModelAddress: MessageListViewModel@5eb7a75
ViewModelAddress: AttachmentsPickerViewModel@10fa0a
ViewModelAddress: MessageComposerViewModel@f90a898
```

Note how the addresses of the VMs are the same in both the activity and the compose component.

_Provide a patch below if it is necessary for testing_

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/channels/ChannelsScreen.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/channels/ChannelsScreen.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/channels/ChannelsScreen.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/channels/ChannelsScreen.kt	(revision 5cc3a84a4a39a26ff8e1ee6ea478c3c72344c585)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/channels/ChannelsScreen.kt	(date 1663761764814)
@@ -16,6 +16,7 @@
 
 package io.getstream.chat.android.compose.ui.channels
 
+import android.util.Log
 import androidx.activity.compose.BackHandler
 import androidx.compose.animation.AnimatedVisibility
 import androidx.compose.animation.ExperimentalAnimationApi
@@ -94,6 +95,8 @@
         factory = viewModelFactory
     )
 
+    Log.d("ViewModelAddress", "Channels: $listViewModel")
+
     val selectedChannel by listViewModel.selectedChannel
     val user by listViewModel.user.collectAsState()
     val connectionState by listViewModel.connectionState.collectAsState()
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/MessagesScreen.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/MessagesScreen.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/MessagesScreen.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/MessagesScreen.kt	(revision 5cc3a84a4a39a26ff8e1ee6ea478c3c72344c585)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/MessagesScreen.kt	(date 1663761764822)
@@ -16,6 +16,7 @@
 
 package io.getstream.chat.android.compose.ui.messages
 
+import android.util.Log
 import androidx.activity.compose.BackHandler
 import androidx.compose.animation.AnimatedVisibility
 import androidx.compose.animation.ExperimentalAnimationApi
@@ -112,6 +113,10 @@
     val attachmentsPickerViewModel =
         viewModel(AttachmentsPickerViewModel::class.java, factory = viewModelFactory)
 
+    Log.d("ViewModelAddress", "Messages (List): $listViewModel")
+    Log.d("ViewModelAddress", "Messages (List): $attachmentsPickerViewModel")
+    Log.d("ViewModelAddress", "Messages (List): $composerViewModel")
+
     val backAction = {
         val isInThread = listViewModel.isInThread
         val isShowingOverlay = listViewModel.isShowingOverlay
Index: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
--- a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(revision 5cc3a84a4a39a26ff8e1ee6ea478c3c72344c585)
+++ b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(date 1663761764809)
@@ -19,6 +19,7 @@
 import android.content.Context
 import android.content.Intent
 import android.os.Bundle
+import android.util.Log
 import androidx.activity.compose.setContent
 import androidx.activity.viewModels
 import androidx.compose.foundation.ExperimentalFoundationApi
@@ -91,6 +92,10 @@
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
+        Log.d("ViewModelAddress", "Messages (List): $listViewModel")
+        Log.d("ViewModelAddress", "Messages (List): $attachmentsPickerViewModel")
+        Log.d("ViewModelAddress", "Messages (List): $composerViewModel")
+
         setContent {
             ChatTheme(dateFormatter = ChatApp.dateFormatter) {
                 MessagesScreen(
Index: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt
--- a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt	(revision 5cc3a84a4a39a26ff8e1ee6ea478c3c72344c585)
+++ b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt	(date 1663761764818)
@@ -19,6 +19,7 @@
 import android.content.Context
 import android.content.Intent
 import android.os.Bundle
+import android.util.Log
 import androidx.activity.compose.setContent
 import androidx.activity.viewModels
 import androidx.compose.foundation.background
@@ -87,6 +88,7 @@
          * You can use the default [ChannelsScreen] component that sets everything up for you,
          * or build a custom component yourself, like [MyCustomUi].
          */
+        Log.d("ViewModelAddress", "Channels: $listViewModel")
         setContent {
             ChatTheme(dateFormatter = ChatApp.dateFormatter) {
                 ChannelsScreen(
```

</details>

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

**Comment by filbabic** [comment]
@adasiewiczr Let me know if this works as an improvement for the issue you raised (when testing).

File: docusaurus/docs/Android/04-compose/05-message-components/01-messages-screen.mdx [hunk 1]
```diff
@@ -70,31 +75,66 @@ Given that the `MessagesScreen` is a screen level solution, it offers limited cu
...
+Here's an example:
+
+```kotlin
+class MessagesActivity : BaseConnectedActivity() {
+
+    // 1
+    private val factory by lazy {
+        MessagesViewModelFactory(
+            context = this,
+            channelId = intent.getStringExtra(KEY_CHANNEL_ID) ?: "",
```
Comment (adasiewiczr) [inline docusaurus/docs/Android/04-compose/05-message-components/01-messages-screen.mdx:106]: Does it make sense to provide an empty cid? Maybe we should suggest `requireNotNull()` instead? 🤔

**Review by adasiewiczr — APPROVED** [review]
LGTM

File: docusaurus/docs/Android/04-compose/05-message-components/01-messages-screen.mdx (hunk 1 above)
Comment (filbabic) [reply docusaurus/docs/Android/04-compose/05-message-components/01-messages-screen.mdx:106]: Yeah definitely makes sense!

**Pull request merged**